src/collected_output.rs

```diff
@@ -1,28 +1,38 @@
 use crate::{Config, Context};
 use std::{
     io::{self, Read, Write},
-    process::{ChildStderr, ChildStdout},
+    process::{ChildStderr, ChildStdin, ChildStdout},
     thread::{self, JoinHandle},
 };
 
 pub(crate) struct Waiter {
+    stdin: JoinHandle<io::Result<()>>,
     stdout: JoinHandle<io::Result<Vec<u8>>>,
     stderr: JoinHandle<io::Result<Vec<u8>>>,
 }
 
 impl Waiter {
     pub(crate) fn spawn_standard_stream_relaying<Stdout, Stderr>(
         context: &Context<Stdout, Stderr>,
-        config: Config,
+        mut config: Config,
+        mut child_stdin: ChildStdin,
         mut child_stdout: ChildStdout,
         mut child_stderr: ChildStderr,
     ) -> Self
     where
         Stdout: Write + Send + Clone + 'static,
         Stderr: Write + Send + Clone + 'static,
     {
+        let mut config_stdin = Vec::new();
+        std::mem::swap(&mut config.stdin, &mut config_stdin);
+        let stdin_join_handle = thread::spawn(move || {
```

Comment:
Does this use a background thread to avoid deadlocking?

Reply:
Yes, exactly. This has to be executed concurrently with the main thread that will spawn the child process. But this also has to be executed concurrently with the other threads in here, that collect `stderr` and `stdout`. Since a child process could potentially do things like write to `stdout` only after having received something on `stdin` (which is pretty normal unix filter behavior) or read from `stdin` only after being able to write to `stdout` (which is pretty weird, but possible). There's also OS buffers involved, but that of course only helps with deadlocks as long as they aren't full.

Of course I wish there were a more elegant and more light-weight way to deal with this than OS threads. But I don't see an obvious solution that is easy to implement and avoids threads.

Reply:
Makes sense. I think this is totally fine. After all, a process is being launched, which is heavier weight than a single thread, so a process + a thread won't be much slower than just a process.